### PR TITLE
Fix: TRANSFER transaction signing

### LINKED ISSUE #16: Invalid signature when sending TRANSFER transaction
It appears `TRANSFER` transactions still don't work via the Java driver. Although the schema seems to match the Python driver perfectly, for both `CREATE` and `TRANSFER`, only `CREATE` works. Digging into BigchainDB's mainline code, it seems like the failure happens [here, in the transaction commons, during validation](https://github.com/bigchaindb/bigchaindb/blob/master/bigchaindb/common/transaction.py#L1048).

In my experiments with the Python driver, `TRANSFER` transactions worked fine. So there is something weird going on with Java. I try to build my transaction like so:

```java
// prepare input
final FulFill sourceXfer = new FulFill();
sourceXfer.setTransactionId(createTxn.getId());
sourceXfer.setOutputIndex(0);

// build asset reference
final String assetId = createTxn.getId();

    // prepare a xfer transaction, and sign it
final BigchainDbTransactionBuilder.ITransactionAttributes attrs = BigchainDbTransactionBuilder
    .init()
    .addInput(null, sourceXfer, (EdDSAPublicKey) owner.getKeypair().getPublic())  // owner is the source identity
    .addOutput(amount, (EdDSAPublicKey) target.getKeypair().getPublic())  // target is the destination identity
    .addAssets(assetId, String.class)
    .operation(Operations.TRANSFER);

// attach metadata, if we have it
if (asset.getMetadata() != null)
  attrs.addMetaData(asset.getMetadata());

final Transaction txn = attrs.buildAndSignOnly(
    (EdDSAPublicKey) owner.getKeypair().getPublic(),
    (EdDSAPrivateKey) owner.getKeypair().getPrivate());
```

But, here is what I get from Bigchain in my `docker-compose` logs:
```
mongodb_1     | <timestamp> I NETWORK  [listener] connection accepted from x.x.x.x:yy #14 (10 connections now open)
mongodb_1     | <timestamp> I NETWORK  [conn14] received client metadata from x.x.x.x:yy conn14: { driver: { name: "PyMongo", version: "3.7.1" }, os: { type: "Linux", name: "Linux", architecture: "x86_64", version: "4.9.93-linuxkit-aufs" }, platform: "CPython 3.6.6.final.0" }

bigchaindb_1  | <timestamp> | [ERROR] (bigchaindb.web.views.base) HTTP API error: 400 - POST:/api/v1/transactions - Invalid transaction, could not validate (InvalidSignature): Transaction signature is invalid. (bigchaindb_webapi - pid: 31)
```

Most notably, of course, is the last line there, which says:
`Invalid transaction, could not validate (InvalidSignature): Transaction signature is invalid.`

This builds on the original bug filed by @NikitaLos (#2, *Transaction operation issue*).

### PR DESCRIPTION
**Pull request by sgammon**
This changeset fixes and closes #16, wherein TRANSFER transactions do not generate valid signatures in the Java driver for BigchainDB. It builds upon the work done in #12 and #14, and fixes the root cause behind @NikitaLos's #10.

After digging into the docs for the Python driver, which *does* sign things correctly, it appeared to do one more step for TRANSFER transactions.

In [this doc page](http://docs.bigchaindb.com/projects/py-driver/en/latest/handcraft.html#id10), it demonstrates that the preimage for the fulfillment digest includes a concatenated postfix, which consists of, in order, and with no separator:
- the transaction ID being fulfilled by an input
- the output index being fulfilled by an input

Consequently, the Bigchain core code seems to [verify signatures the same way](https://github.com/bigchaindb/bigchaindb/blob/master/bigchaindb/common/transaction.py#L1040). However, the Java driver was not considering these factors, instead [digesting transactions the same way regardless of type](https://github.com/bigchaindb/java-bigchaindb-driver/blob/master/src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java#L379).

Changes in this PR:
- [x] Apply a fix, by appending the requisite items before hashing
  - [x] Iterate over all fulfilled outputs (inputs) and append them
- [x] Update project readme to include sample of `TRANSFER` operation

### CODE REVIEW

File: src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java [hunk 1]
```diff
@@ -371,12 +371,26 @@ public IBuild build(EdDSAPublicKey publicKey) {
...
-            log.debug("TO BE HASHED ---->\n" + temp + "\n<");
             JsonObject transactionJObject = DriverUtils.makeSelfSortingGson(temp);
 
-            byte[] sha3Hash = DriverUtils.getSha3HashRaw(transactionJObject.toString().getBytes());
+            byte[] sha3Hash;
+            if (Operations.TRANSFER.name().equals(this.transaction.getOperation())) {
+                // it's a transfer operation: make sure to update the hash pre-image with
+                // the fulfilling transaction IDs and output indexes
+                StringBuilder preimage = new StringBuilder(transactionJObject.toString());
+                for (Input in : this.transaction.getInputs()) {
```
Comment (sgammon) [inline src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java:382]: the fix itself: if it is a `TRANSFER` operation, do the following:
- `for` each `Input`...
  - concatenate `transactionId` and `String.valueOf(outputIndex)`
  - add to end of hash preimage

all before calculating the hash and signing.

File: README.md [hunk 2]
```diff
@@ -72,56 +72,78 @@ BigchainDbConfigBuilder
...
-
+	.buildAndSign((EdDSAPublicKey) keyPair.getPublic(), (EdDSAPrivateKey) keyPair.getPrivate())
+	.sendTransaction();
 ```
 
-### Example: Create, Sign and Send a Transaction
+### Example: Transfer an Asset
+
+Performing a TRANSFER transaction in BigchainDB changes an asset's ownership (or, more accurately, authorized signers):
+
```
Comment (sgammon) [inline README.md:109]: add an example to the readme of a `TRANSFER` operation. this would have really helped me get started.

**Review by innopreneur — APPROVED** [review]
looks good

**Comment by ghost** [comment]
Thanks @sgammon!

**Pull request merged**